171 chain with ignored fields

NRBPerdijk opened the pull request:
# What problem does this pull request solve?
Fixes issue #171 where the user couldn't chain multiple calls to withOnlyTheseFields and withIgnoredFields and an incorrect error message would appear

# Please provide any additional information below.
Some mild refactoring to optimise the adding of ignored and expected fields.
Important: there are no warnings to signify the user has specified the same field twice (works as before).
Rest of the information on what has changed and why should be extractable from the commit messages. 👍

NRBPerdijk committed:
added unittests specifying the desired behaviour: the user should be able to chain calls of withOnlyTheseFields OR withIgnoredFIelds, but not be able to combine both. Currently all three scenarios result in an error: You can call either withOnlyTheseFields or withIgnoredFields, but not both. Unittests specifying the desired behavior of chaining currently fail.

NRBPerdijk committed:
missed a line in the unittest: the given class has two fields that should be used, so added the second field to the include clause

NRBPerdijk committed:
Created the FieldNameExtractor, a helper lass with a static method that extracts all the fieldnames of a given classtype and returns them as a Set<String>

NRBPerdijk committed:
Created tests for FieldNameExtractor

NRBPerdijk committed:
made FieldNameExtractor return an UnmodifiableSet, validated this with extra unittests

NRBPerdijk committed:
Added actualFields to Configuration

NRBPerdijk committed:
Replaced the build-up for actual fields in withOnlyTheseFields with a call to Configuration.getActualFields() to avoid repeated reflection.

NRBPerdijk committed:
Replaced the build-up for actual fields in validateFieldNamesExist with call to config.getActualFields(), replaced the validation in withOnlyTheseFields() with call to validateFieldNamesExist.

NRBPerdijk committed:
Added includedFields to enable tracking of which with method was called (to allow chaining of the same method). Renamed ignoredFields to excludedFields to illuminate that these two belong together, kept the withIgnoredField method, because it will host logic later, currently it returns a copy of excludedFields.

NRBPerdijk committed:
This is the core change:
withOnlyTheseFields and withIgnoredFields now check whether the other method has already been called, if so, they throw an exception.
both methods no longer check whether they themselves have been called, but instead add the content of any new call to the already existing collection.
getIgnoredFields now returns the excludedFields, unless excludedFields is empty AND includedFields is NOT empty; if the latter is the case, it returns a list of all the actualFields that do not occur in includedFields.

NRBPerdijk committed:
Renamed extractFields to extractFieldNames because that is what the method actually does.
Fixed compiler warning concerning generics by making the static method use type parameters.

NRBPerdijk committed:
Fixed compiler checkstyle warnings due to codeformatting rules, the presence of licensing etc.

NRBPerdijk committed:
Changed copyright on the new files to the name of the file-creator (myself).

jqno commented on src/test/java/nl/jqno/equalsverifier/integration/extended_contract/SignificantFieldsTest.java:
```diff
@@ -163,6 +163,31 @@ public void succeed_whenTwoFieldsAreUnused_givenAllFieldsShouldBeUsedExceptThose
     }
 
     @Test
+    public void succeed_repeating_withIgnoredFields_givenAllFieldsShouldBeUsedExceptThoseTwo() {
```
The integration tests have a convention, where the name of each test case is a sort of reverse Given/When/Then. The clauses are separated by `_` for readability, but there should be no `_`s in other spots.

jqno commented on src/main/java/nl/jqno/equalsverifier/EqualsVerifier.java:
```diff
@@ -334,17 +324,20 @@ private void checkNonnullFields() {
-        Set<String> actualFieldNames = new HashSet<>();
-        for (Field field : FieldIterable.of(config.getType())) {
-            actualFieldNames.add(field.getName());
+    private void checkNoExistingIncludedFields() {
```
This method and the previous one are nearly identical. Let's merge them! :)

jqno commented on src/main/java/nl/jqno/equalsverifier/internal/util/Configuration.java:
```diff
@@ -79,90 +84,126 @@ public PrefabValues getPrefabValues() {
     public Set<String> getIgnoredFields() {
-        return Collections.unmodifiableSet(ignoredFields);
+        Set<String> ignoredFields = new HashSet<>();
+        if (excludedFields.isEmpty() && !includedFields.isEmpty()) {
```
Let's reduce cyclomatic complexity by inverting the condition and returning the `ignoredFields` immediately, and then de-indent the for loop.

jqno reviewed (commented):
Thank you, this is a very nice PR!
I have a couple of points but they're very minor.

NRBPerdijk committed:
Rename of tests to conform to project standard.

Processed review comment: The newly introduced tests in SignificantFieldsTest now conform more strictly to the pattern the other tests follow.

NRBPerdijk committed:
Combined NoExisting methods and renamed them.

Processed review comment: the checkNoExistingExcludedFields and checkNoExistingExcludedFields now have their shared logic extracted to a third method. The naming has also been adjusted: check has been replaced with assert to make more expressive that this is an expected state of the application that will result in a thrown exception if it is not true. Also applied this change to checkNoNonnullFields to be consistent.

NRBPerdijk committed:
Reduced complexity in getIgnoredFields()

Processed review comment: inverted and simplified the boolean condition to only check if includedFields is empty. If it is, excludedFields should always be the result (even if it is empty, because then no fields are ignored), otherwise it composes ignoredFields by returning the actualFields without the field names that occur in includedFields.
